### ISSUE TITLE
`enable_grad` context doesn't work as expected in backward function of torch.autograd.Function

### ISSUE THREAD

**Issue by lxuechen** [pytorch/pytorch#31715]
## 🐛 Bug

This hasn't been an issue before, but it seems right now, it could be a serious blocking issue of one of my research projects. The gist is that a call to `torch.autograd.grad` seems to disable the `enable_grad` context in the `backward` function of `torch.autograd.Function`.

## To Reproduce
Here's a mini-repro:
```Python
import torch
from torch import nn


class BackwardBreaks(torch.autograd.Function):

    @staticmethod
    def forward(ctx, *args, **kwargs):
        assert len(args) == 3
        model = args[0]
        inp1, inp2 = args[1], args[2]
        ctx.model = model
        ctx.save_for_backward(inp1, inp2)
        return model.f(inp1) + model.g(inp2)

    @staticmethod
    def backward(ctx, *grad_outputs):
        model, inps = ctx.model, ctx.saved_tensors
        inp1, inp2 = inps
        grad_outputs, = grad_outputs
        with torch.enable_grad():
            print(torch._C.is_grad_enabled())
            grad1, = torch.autograd.grad(
                inp1 * 3, inp1, grad_outputs=grad_outputs
            )
            print(torch._C.is_grad_enabled())
            grad2, = torch.autograd.grad(
                inp2 ** 2, inp2, grad_outputs=grad_outputs
            )
        return None, grad1, grad2


class Model(nn.Module):
    def __init__(self):
        super().__init__()

    def f(self, inp):
        return inp * 2.

    def g(self, inp):
        return inp ** 2.


def backward_breaks_wrapper(model, inp1, inp2):
    return BackwardBreaks.apply(model, inp1, inp2)


def main():
    model = Model()
    inp1 = torch.randn(3, 1, requires_grad=True)
    inp2 = torch.randn(3, 1, requires_grad=True)

    result = backward_breaks_wrapper(model, inp1, inp2)
    result.backward(torch.ones(3, 1))
    print(inp1.grad)
    print(inp2.grad)


if __name__ == '__main__':
    main()
```

Run this w/ torch==1.3.0. I didn't try the bleeding edge version.

## Expected behavior
Clearly, we don't expect the output of `print(torch._C.is_grad_enabled())` to change. But afaik, the actual outputs here are `True` and then `False`.

## Environment
```
Collecting environment information...
PyTorch version: 1.3.1
Is debug build: No
CUDA used to build PyTorch: 10.1.243

OS: Debian GNU/Linux rodete
GCC version: (Debian 8.3.0-6) 8.3.0
CMake version: version 3.13.4

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: GeForce GTX 1080
GPU 1: Quadro P1000

Nvidia driver version: 430.50
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] numpy==1.17.0
[pip] torch==1.3.1
[pip] torchdiffeq==0.0.1
[pip] torchfile==0.1.0
[pip] torchvision==0.4.2
[conda] torch                     1.3.1                    pypi_0    pypi
[conda] torchdiffeq               0.0.1                    pypi_0    pypi
[conda] torchfile                 0.1.0                    pypi_0    pypi
[conda] torchsnooper              0.7.1                    pypi_0    pypi
[conda] torchvision               0.4.2                    pypi_0    pypi
```

## Additionals
Clearly, we could bypass this issue by setting `create_graph=True` in the first call to `torch.autograd.grad`. However, the other implications of this flag here isn't that great, considering that we're not aiming to obtain high-order derivs. 

Alternatively, we could use a new `torch.enable_grad` scope each time we decide to use `torch.autograd.grad`. Though this approach is less clean, and it's still nice to figure out why a call to `grad` somehow disable the context. Note, this doesn't happen if it's not in the backward function. 

Moreover, there seems to other bugs (which I haven't stably been able to get a mini-repro) in this backward function that's causing some non-deterministic segfaults and runtime errors even after I fix both the seed of torch and numpy. 

cc @ezyang @SsnL @albanD @zou3519 @gqchen

**Comment by lxuechen** [comment]
Any updates?

**Comment by albanD** [comment]
Thanks for the report, I will take a look why this happens. It might simply be that we don't restore the status in `autograd.grad` properly.

**Comment by lxuechen** [comment]
I was also able to escape another nondeterministic (potentially concurrency) bug by replacing all retain graph with create graph.

**Comment by albanD** [comment]
So the problem is [here](https://github.com/pytorch/pytorch/blob/1ba1799a66b20f17657a132eab111d002b3356f3/torch/csrc/autograd/engine.cpp#L314) where we don't restore the state properly.
I'll send a PR to fix this. The workaround of using `create_graph=True` will work for now.

**Comment by lxuechen** [comment]
Thanks for the patch! This one seems so subtle.

**Comment by albanD** [comment]
Yes, it is fairly sneaky, it only happens when you use re-entrant autograd (backward within a backward) and you modify the grad-mode by hand inside, and reuse it multiple times.
